### ISSUE TITLE
Inlay hint colorings

### ISSUE THREAD

**Issue by Ciantic** [rust-analyzer/rust-analyzer#5337]
Suggestion: Have configurable inlay hint colorings. At least two: if the inlay hint pushes the text, it may have different color. If the inlay hint is at the end of line, have a different color option.

I've noticed a pattern: when the inlay hints comes at the *end* of the line, I want them to be more grayed out. But when they push out the text further, I'd like them to be way more visible.

Here is a classic case of this: https://github.com/rust-analyzer/rust-analyzer/issues/4079

Notice how the image:

> ![Bad indentation](https://user-images.githubusercontent.com/1502855/79914068-e8582000-8424-11ea-96e8-5ba04d2e168f.png)

Makes it looks like the indendation is wrong. This is because the inlay hints looks as if they are empty space, because they are so invisible. 

Yet there is nothing wrong with the way inlay hints are in there, or with their indentation, it's just illusion because the color is so invisible.

If the inlay hints that push the text could be more visible it would make the indentation to look right IMO.

**Comment by SomeoneToIgnore** [comment]
rust-analyzer server only provides the hint text and type, and the decorations are implemented on the client side.

On VSCode client side, we separate the hints into 3 different categories (type, param and chaining hints), each having a separate style config.
All the code related to the styling is located in the [inlay_hints.ts](https://github.com/rust-analyzer/rust-analyzer/blob/5ca7cd9/editors/code/src/inlay_hints.ts) file, here's the type hints decoration options:

https://github.com/rust-analyzer/rust-analyzer/blob/5ca7cd9/editors/code/src/inlay_hints.ts#L43-L57

From the VSCode API point of view, the decoration styling boids down to the options defined in https://vshaxe.github.io/vscode-extern/vscode/ThemableDecorationAttachmentRenderOptions.html
For the issue ticket, `color` property can be used.

In order to add the color, you need to add more options (with reasonable defaults for both light and dark themes) somewhere here: https://github.com/rust-analyzer/rust-analyzer/blob/5ca7cd9/editors/code/src/config.ts#L99-L107 and then use it in `inlay_hints.ts`.

**Comment by Veetaha** [comment]
It is way simpler than you think. This way the user can control the color of the hints in `settings.json` now:
```json
    "workbench.colorCustomizations": {
        "[Default Dark+]": {
            "rust_analyzer.inlayHint": "#a70000",
        }
    },
```
Currently, this is the most granular way to do this, but it is simple to make it more configurable:
Add two more colors configs to this array:
https://github.com/rust-analyzer/rust-analyzer/blob/4b1c3724364ff538acdcb44bdd23d501ef54cff3/editors/code/package.json#L686-L694
Say `rust_analyzer.inlayHint.parameterHints`, `rust_analyzer.inlayHint.typeHints` and `rust_analyzer.inlayHint.chainingHints` (I recommend being consistent in naming with `rustlanalyzer.inlayHints.*` settings, though this is currently already inconsistent :D).

Change the `new vscode.ThemeColor('rust_analyzer.inlayHint')` line to use appropriate `rust_analyzer.inlayHint.*` id from `package.json` here:

https://github.com/rust-analyzer/rust-analyzer/blob/4b1c3724364ff538acdcb44bdd23d501ef54cff3/editors/code/src/inlay_hints.ts#L43-L89

Probably add some user docs for this feature

**Comment by Ciantic** [comment]
@Veetaha my gripe is with single color. I want multiple colors. Perhaps starting with giving typeHints, chainingHints, parameterHints a different color.

(Also I noticed something else, I would like to hide parameter name hints for single argument functions, or two argument functions with different types if they are on single line.)

**Comment by tmcguire** [comment]
On a related note, I wish the background color of the inlay hint was configurable. With a background color, it would be much easier to distinguish inlay hints from the surrounding code.
Example from Android Studio's default configuration: 
![image](https://user-images.githubusercontent.com/1164876/91178794-f5905e80-e6e5-11ea-846d-1c8856c147ad.png)

**Comment by Veetaha** [comment]
Vscode doesn't provide an API to configure the font size of our inlay hints but changing the background color is doable (#5885), otherwise I suggest pushing the upstream issue forward: https://github.com/microsoft/language-server-protocol/issues/956

**Comment by Veykril** [comment]
Configuration for this was actually implemented in #5885, background coloring as said is relient on the client offering such a feature(and judging from the issue it will never happen unfortuantely) so I'll go ahead and close this.

**Comment by boogiefromzk** [comment]
For ones who wonder (replace the theme with your preferred one), edit settings.json adding:
```
    "workbench.colorCustomizations": {
        "[Default Light+]": {
            "editorInlayHint.background": "#ffffff",
            "editorInlayHint.foreground": "#aaaaaa"
        }
    }
```
